### PR TITLE
DT-727 Add Spring Boot info and manifest attributes

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPlugin.kt [hunk 1]
```diff
@@ -6,19 +6,33 @@ import org.gradle.api.Plugin
...
     applyRepositories(project)
     applyDependencyManagementBom(project)
+    setSpringBootInfo(project)
+    setManifestAttributes(project)
     addDependencies(project)
     setKotlinCompileJvmVersion(project)
   }
 
+  private fun getVersion(): String {
+    return if (System.getenv().contains("BUILD_NUMBER")) System.getenv("BUILD_NUMBER")
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPlugin.kt:32]: I coudn't think of a nice way to test this without adding a load of complexity due to the System.getenv calls. I don't think it's a huge problem but any ideas appreciated.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPlugin.kt [hunk 2]
```diff
@@ -53,4 +67,29 @@ class DpsSpringBootPlugin : Plugin<Project> {
...
+  private fun setSpringBootInfo(project: Project) {
+    val sbExtension = project.extensions.getByName("springBoot") as SpringBootExtension
+    sbExtension.buildInfo { buildInfo ->
+      buildInfo.properties { buildInfoProps ->
+        buildInfoProps.time = Instant.now()
+        buildInfoProps.additional = getAdditionalBuildInfo()
+        buildInfoProps.version = project.version.toString()
+      }
+    }
+  }
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPlugin.kt:80]: Sets Spring Boot info as done in the `springBoot {` calls in gradle builds.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPlugin.kt [hunk 3]
```diff
@@ -53,4 +67,29 @@ class DpsSpringBootPlugin : Plugin<Project> {
...
+        "by" to System.getProperty("user.name"),
+        "operatingSystem" to "${System.getProperty("os.name")} (${System.getProperty("os.version")})",
+        "machine" to InetAddress.getLocalHost().hostName)
+  }
+
+  private fun setManifestAttributes(project: Project) {
+    val manifest = (project.tasks.getByName("bootJar") as BootJar).manifest
+    manifest.attributes["Implementation-Version"] = project.version
+    manifest.attributes["Implementation-Title"] = project.name
+  }
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPlugin.kt:93]: Sets the manifest attributes as done in the ```
tasks {
  bootJar { ...
```
calls in our gradle builds.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPluginTest.kt [hunk 4]
```diff
@@ -105,4 +109,24 @@ class DpsSpringBootPluginTest {
...
+  @Test
+  fun `Should set build info`() {
+    val properties = (project.tasks.getByPath("bootBuildInfo") as BuildInfo).properties
+    assertThat(properties.additional).extracting("by").isEqualTo(System.getProperty("user.name"))
+    assertThat(properties.additional).extracting("operatingSystem").isNotNull()
+    assertThat(properties.additional).extracting("machine").isNotNull()
+
+    assertThat(properties.time).isNotNull()
+
+    assertThat(properties.version).isEqualTo(LocalDate.now().format(ISO_DATE))
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPluginTest.kt:122]: is this because we have no build number?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPluginTest.kt [hunk 5]
```diff
@@ -105,4 +109,24 @@ class DpsSpringBootPluginTest {
...
   }
+
+  @Test
+  fun `Should set build info`() {
+    val properties = (project.tasks.getByPath("bootBuildInfo") as BuildInfo).properties
+    assertThat(properties.additional).extracting("by").isEqualTo(System.getProperty("user.name"))
+    assertThat(properties.additional).extracting("operatingSystem").isNotNull()
+    assertThat(properties.additional).extracting("machine").isNotNull()
+
+    assertThat(properties.time).isNotNull()
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPluginTest.kt:120]: maybe get the time, parse and check the date isToday() ?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/KotlinFuncTest.kt [hunk 6]
```diff
@@ -30,6 +34,14 @@ class KotlinFuncTest {
     assertThat(webVersion).isEqualTo(actuatorVersion)
   }
 
+  @Test
+  fun `Manifest file contains project name and version`() {
+    val file = findJar(projectDir, "spring-boot-project-kotlin")
+    val jarFile = JarFile(file)
+    assertThat(jarFile.manifest.mainAttributes.getValue("Implementation-Version")).isEqualTo(LocalDateTime.now().format(DateTimeFormatter.ISO_DATE))
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/KotlinFuncTest.kt:41]: surprised this works, wouldn't time moved forward a little bit?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPluginTest.kt (hunk 4 above)
Comment (mikehalmamoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPluginTest.kt:122]: Yes... locally it defaults to today, but on Circle it uses the `BUILD_NUMBER` env var.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPluginTest.kt (hunk 5 above)
Comment (mikehalmamoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/DpsSpringBootPluginTest.kt:120]: I was going to do that, then thought about the edge case where the build starts before midnight but the test asserts after it.  Which is silly.

I'll do it in the next PR.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/KotlinFuncTest.kt (hunk 6 above)
Comment (mikehalmamoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/gradle/KotlinFuncTest.kt:41]: It's just a date not a date time - I'll use LocalDate in the next PR.
